test/realizations/catchments/Bmi_Cpp_Adapter_Test.cpp

```diff
@@ -0,0 +1,664 @@
+    std::vector<std::string> expected_output_var_locations = { "node", "node" };
+    std::vector<int> expected_output_var_grids = { 0, 0 };
+    std::vector<std::string> expected_output_var_units = { "m", "m" };
+    std::vector<std::string> expected_output_var_types = { "double", "double" };
+    int expected_grid_rank = 1;
+    int expected_grid_size = 1;
+    std::string expected_grid_type = "scalar";
+    int expected_var_nbytes = 8; //type double
+
+};
+
+void Bmi_Cpp_Adapter_Test::SetUp() {
+    // Uses the same config files as the C test model...
+    std::vector<std::string> config_path_options = {
+            "test/data/bmi/test_bmi_c/",
+            "./test/data/bmi/test_bmi_c/",
+            "../test/data/bmi/test_bmi_c/",
+            "../../test/data/bmi/test_bmi_c/",
+    };
+    std::string config_basename_0 = "test_bmi_c_config_0.txt";
+    config_file_name_0 = file_search(config_path_options, config_basename_0);
+
+    std::vector<std::string> forcing_dir_opts = {"./data/forcing/", "../data/forcing/", "../../data/forcing/"};
+    forcing_file_name_0 = file_search(forcing_dir_opts, "cat-27_2015-12-01 00_00_00_2015-12-30 23_00_00.csv");
+
+    std::vector<std::string> lib_dir_opts = {
+            "./extern/test_bmi_cpp/cmake_build/",
+            "../extern/test_bmi_cpp/cmake_build/",
+            "../../extern/test_bmi_cpp/cmake_build/"
+    };
+    lib_file_name_0 = file_search(lib_dir_opts, BMI_TEST_CPP_LOCAL_LIB_NAME);
+    bmi_module_type_name_0 = "test_bmi_cpp";
+    try {
+        adapter = std::make_unique<Bmi_Cpp_Adapter>(bmi_module_type_name_0, lib_file_name_0, config_file_name_0, 
+                                                forcing_file_name_0, false, true, CREATOR_FUNC, DESTROYER_FUNC,
+                                                utils::StreamHandler());
+    }
+    catch (const std::exception &e) {
+        std::clog << e.what() << std::endl;
+        throw e;
+    }
+ }
+
+void Bmi_Cpp_Adapter_Test::TearDown() {
+
+}
+
+std::string
+Bmi_Cpp_Adapter_Test::file_search(const std::vector<std::string> &parent_dir_options, const std::string& file_basename) {
+    // Build vector of names by building combinations of the path and basename options
+    std::vector<std::string> name_combinations;
+
+    // Build so that all path names are tried for given basename before trying a different basename option
+    for (auto & path_option : parent_dir_options)
+        name_combinations.push_back(path_option + file_basename);
+
+    return utils::FileChecker::find_first_readable(name_combinations);
+}
+
+/** Simple test to make sure the model initializes. */
+TEST_F(Bmi_Cpp_Adapter_Test, Initialize_0_a) {
+    adapter->Initialize();
+    adapter->Finalize();
+}
+
+/** Test output variables can be retrieved. */
+TEST_F(Bmi_Cpp_Adapter_Test, GetOutputVarNames_0_a) {
+    try {
+        ASSERT_EQ(adapter->GetOutputVarNames(), expected_output_var_names);
+    }
+    catch (std::exception& e) {
```

Comment:
Can't `adapter->GetOutputVarNames()` throw a `runtime_error` if the BMI model returns `BMI_FAILURE` for whatever reason?  At least I thought that was what most of the proxy functions did...maybe not???

Reply:
Ah!...but the C++ version of `GetOutputVarNames()` doesn't return a success or failure value, it returns a vector.
https://github.com/NOAA-OWP/ngen/blob/9e42219a12e5fb756e7fc08dfdf4dfc48bb87d81/include/bmi.hpp#L44

Reply:
Ok, re-direct.  Do we expect `BMI::GetOutputVarNames()` to throw an exception that mirrors the `BMI_FAILURE` concept?

Reply:
From the [BMI docs](https://bmi-spec.readthedocs.io/en/latest/#:~:text=in%20these%20languages.-,In%20C%20and%20Fortran%2C%20an%20integer%20status%20code%20indicating%20success%20or%20failure%20is%20returned.%20In%20C%2B%2B%20and%20Python%2C%20an%20exception%20is%20raised%20on%20failure.,-%5BModel%20control%20functions).

Wouldn't it be nice if they told us what exception to expect?  Maybe even wrote an exception class to go along with BMI?  If only I had a place to put such things...

Reply:
I think so? To my knowledge BMI doesn't enumerate the possible exceptions that can be thrown, so the best we can do otherwise is probably just handle any exception as `BMI_FAILURE`...it's as good as you can do in C (not like you can respond to a boolean failure code with any more granularity).

Reply:
For now...